### ISSUE TITLE
Bootstrap 4 

### ISSUE THREAD

**Issue by drjoju** [nWidart/laravel-menus#13]
Hi,
https://blackrockdigital.github.io/startbootstrap-modern-business/
As you can see in the Bootstrap 4 menu, the first lavel of link has the class "nav-link". For example 

`<a class="nav-link" href="about.html">About</a>
`
In the second level the class is different, For example.
`
<a class="dropdown-item" href="portfolio-3-col.html">3 Column Portfolio</a>
`

As I see in the presenter there is only one place to set de class

```
    public function getMenuWithoutDropdownWrapper($item)
    {
        return '<li class="nav-item" ' . $this->getActiveState($item) . '><a class="nav-link" href="' . $item->getUrl() . '" ' . $item->getAttributes() . '>' . $item->getIcon() . ' ' . $item->title . '</a></li>' . PHP_EOL;
    }
```

How can we set the correct class in function of depth?

Thanks and best regards

**Comment by nWidart** [comment]
you can create a new presenter class for bootstrap 4.

**Comment by drjoju** [comment]
Yes, I do it but I don't know how to process the second level. I created a presenter like this one but I don't know if it's the best solution
`
<?php

namespace App\Presenters;

use Nwidart\Menus\MenuItem;
use Nwidart\Menus\Presenters\Presenter;

class FrontMenuNavbarPresenter extends Presenter
{
    /**
     * {@inheritdoc }.
     */
    public function getOpenTagWrapper()
    {
        return PHP_EOL . '<ul class="navbar-nav ml-auto">' . PHP_EOL;
    }

    /**
     * {@inheritdoc }.
     */
    public function getCloseTagWrapper()
    {
        return PHP_EOL . '</ul>' . PHP_EOL;
    }

    /**
     * {@inheritdoc }.
     */
    public function getMenuWithoutDropdownWrapper($item)
    {
        return '<li class="nav-item" ' . $this->getActiveState($item) . '><a class="nav-link" href="' . $item->getUrl() . '" ' . $item->getAttributes() . '>' . $item->getIcon() . ' ' . $item->title . '</a></li>' . PHP_EOL;
    }

    /**
     * {@inheritdoc }.
     */
    public function getSubmenuWithoutDropdownWrapper($item)
    {
        return '<li class="nav-item" ' . $this->getActiveState($item) . '><a class="dropdown-item" href="' . $item->getUrl() . '" ' . $item->getAttributes() . '>' . $item->getIcon() . ' ' . $item->title . '</a></li>' . PHP_EOL;
    }


    /**
     * {@inheritdoc }.
     */
    public function getActiveState($item, $state = ' class="active"')
    {
        return $item->isActive() ? $state : null;
    }

    /**
     * Get active state on child items.
     *
     * @param $item
     * @param string $state
     *
     * @return null|string
     */
    public function getActiveStateOnChild($item, $state = 'active')
    {
        return $item->hasActiveOnChild() ? $state : null;
    }

    /**
     * {@inheritdoc }.
     */
    public function getDividerWrapper()
    {
        return '<li class="divider"></li>';
    }

    /**
     * {@inheritdoc }.
     */
    public function getHeaderWrapper($item)
    {
        return '<li class="dropdown-header">' . $item->title . '</li>';
    }

    /**
     * {@inheritdoc }.
     */
    public function getMenuWithDropDownWrapper($item)
    {
        return '<li class="nav-item dropdown' . $this->getActiveStateOnChild($item, ' active') . '">
		          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown">
					' . $item->getIcon() . ' ' . $item->title . '
			      	<b class="caret"></b>
			      </a>
			      <ul class="dropdown-menu">
			      	' . $this->getChildMenuItems($item) . '
			      </ul>
		      	</li>'
            . PHP_EOL;
    }

    /**
     * Get multilevel menu wrapper.
     *
     * @param \Nwidart\Menus\MenuItem $item
     *
     * @return string`
     */
    public function getMultiLevelDropdownWrapper($item)
    {
        return '<li class="nav-item dropdown' . $this->getActiveStateOnChild($item, ' active') . '">
		          <a href="#" class="dropdown-item dropdown-toggle" data-toggle="dropdown">
					' . $item->getIcon() . ' ' . $item->title . '
			      	<b class="caret pull-right caret-right"></b>
			      </a>
			      <ul class="dropdown-menu">
			      	' . $this->getChildMenuItems($item) . '
			      </ul>
		      	</li>'
            . PHP_EOL;
    }

    /**
     * Get child menu items.
     *
     * @param \Nwidart\Menus\MenuItem $item
     *
     * @return string
     */
    public function getChildMenuItems(MenuItem $item)
    {
        $results = '';
        foreach ($item->getChilds() as $child) {
            if ($child->hidden()) {
                continue;
            }

            if ($child->hasSubMenu()) {
                $results .= $this->getMultiLevelDropdownWrapper($child);
            } elseif ($child->isHeader()) {
                $results .= $this->getHeaderWrapper($child);
            } elseif ($child->isDivider()) {
                $results .= $this->getDividerWrapper();
            } else {
                $results .= $this->getSubmenuWithoutDropdownWrapper($child);
            }
        }

        return $results;
    }
}
`

What do you think?

Thanks

**Comment by nWidart** [comment]
Plus format your post, this is unreadable :D

**Comment by drjoju** [comment]
Sorry,

Basically I created a presenter based on NavbarPresenter and override the method getChildMenuItems.

The part modified is
```

            if ($child->hasSubMenu()) {
                $results .= $this->getMultiLevelDropdownWrapper($child);
            } elseif ($child->isHeader()) {
                $results .= $this->getHeaderWrapper($child);
            } elseif ($child->isDivider()) {
                $results .= $this->getDividerWrapper();
            } else {
                $results .= $this->getSubmenuWithoutDropdownWrapper($child);
            }


```
I introduce the method getSubmenuWithoutDropdownWrapper with this code
```

    public function getSubmenuWithoutDropdownWrapper($item)
    {
        return '<li class="nav-item" ' . $this->getActiveState($item) . '><a class="dropdown-item" href="' . $item->getUrl() . '" ' . $item->getAttributes() . '>' . $item->getIcon() . ' ' . $item->title . '</a></li>' . PHP_EOL;
    }

```
This method write the correct class (dropdown-item) in the submenu.

What do you think about this aproach?

Thanks

**Comment by vpratfr** [comment]
Nevermind. Ended up rewriting your package for a few things I needed. 

Fixed in https://github.com/marvinlabs/laravel-menus/

**Comment by laxxis** [comment]
Is there any solution at this point?

**Comment by asaelmedina** [comment]
i cant make a present with bootstrap 4 :(

**Comment by asaelmedina** [comment]
I did it. I used the idea of @drjoju . My implementation it was with AdminLTE 3

**Comment by asaelmedina** [comment]
[SidebarMenuPresenter.txt](https://github.com/nWidart/laravel-menus/files/4208243/SidebarMenuPresenter.txt)
